File: backend/stress_test.py
import asyncio
import json
import random
import time
import urllib.request
from collections import deque
import websockets
from dotenv import dotenv_values

# Pega o token de admin do .env local
env = dotenv_values(".env")
ADMIN_TOKEN = env.get("ADMIN_TOKEN", "")

BASE_URL = "http://localhost:8000"
WS_URL = "ws://localhost:8000"

categories = [
    {"name": "Bolo de Pote", "price": 10},
    {"name": "Refrigerante 350ml", "price": 5},
    {"name": "Fatia de Pizza", "price": 15},
    {"name": "Pulseira Neon", "price": 8},
    {"name": "Hambúrguer", "price": 25}
]

store_names = [
    "Cantina Central",
    "Barraca de Jogos",
    "Bazar 3º Ano",
    "Doces e Cia",
    "Artesanato Local"
]

active_comandas = []
metrics = {
    "comandas_created": 0,
    "debits_requested": 0,
    "debits_success": 0,
    "debits_failed": 0,  # Geralmente saldo insuficiente
    "errors": 0,
    "latencies": deque(maxlen=200)
}

def post_json(endpoint, data, token):
    req = urllib.request.Request(f"{BASE_URL}{endpoint}", data=json.dumps(data).encode('utf-8'))
    req.add_header('Content-Type', 'application/json')
    req.add_header('token', token)
    try:
        with urllib.request.urlopen(req) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        if e.code != 400: # Ignora 400 se a categoria/loja já existir
            print(f"[{e.code}] Erro ao chamar {endpoint}: {e.read().decode()}")
        return None
    except Exception as e:
        print(f"Erro de conexão com {endpoint}: {e}")
        return None

async def admin_worker():
    uri = f"{WS_URL}/ws/admin?token={ADMIN_TOKEN}"
    while True:
        try:
            async with websockets.connect(uri) as ws:
                print("[🏦 Admin Worker] Conectado e emitindo comandas!")
                names = ["Pedro", "Maria", "João", "Ana", "Lucas", "Julia", "Marcos"]
                
                while True:
                    # Emite comanda com valor inicial aleatório entre 50 e 150 ETC
                    holder = random.choice(names) + f" {random.randint(100, 999)}"
                    balance = random.randint(50, 150)
                    
                    # Simula adição de crédito atrelado à compra de produtos iniciais (ex: Banco vendeu o ingresso e pulseras)
                    cart = [
                        {"name": random.choice(categories)["name"], "quantity": random.randint(1, 3)}
                    ]
                    
                    msg = {
                        "action": "create_comanda",
                        "holder_name": holder,
                        "initial_balance": balance,
                        "cart_items": cart
                    }
                    start_t = time.time()
                    await ws.send(json.dumps(msg))
                    resp = json.loads(await ws.recv())
                    latency = (time.time() - start_t) * 1000
                    metrics["latencies"].append(latency)
                    
                    if resp.get("status") == "success":
                        metrics["comandas_created"] += 1
                        active_comandas.append(resp["data"]["code"])
                    else:
                        metrics["errors"] += 1
                        
                    # Velocidade de emissão (ex: 2 comandas por segundo = 0.5s de taxa)
                    await asyncio.sleep(0.5) 
        except Exception as e:
            print(f"[🏦 Admin Erro] {e}. Reconectando...")
            metrics["errors"] += 1
            await asyncio.sleep(2)

async def store_worker(store_token, store_name):
    uri = f"{WS_URL}/ws/store?token={store_token}"
    while True:
        try:
            async with websockets.connect(uri) as ws:
                print(f"[🏪 Store Worker - {store_name}] Operacional e cobrando!")
                while True:
                    if active_comandas:
                        comanda_code = random.choice(active_comandas)
                        amount = random.randint(5, 30)
                        
                        msg = {
                            "action": "debit",
                            "comanda_code": comanda_code,
                            "amount": amount
                        }
                        metrics["debits_requested"] += 1
                        start_t = time.time()
                        await ws.send(json.dumps(msg))
                        resp = json.loads(await ws.recv())
                        latency = (time.time() - start_t) * 1000
                        metrics["latencies"].append(latency)
                        
                        if resp.get("status") == "success":
                            metrics["debits_success"] += 1
                        else:
                            metrics["debits_failed"] += 1
                            
                    # Simula fluxo maluco: lojas tentando debitar entre 0.1 e 0.6s
                    await asyncio.sleep(random.uniform(0.1, 0.6))
        except Exception as e:
            print(f"[🏪 Store {store_name} Offline] {e}. Reparando máquina de cartão...")
            metrics["errors"] += 1
            await asyncio.sleep(2)

async def reporter_loop():
    start_time = time.time()
    while True:
        await asyncio.sleep(3)
        elapsed = time.time() - start_time
        total_ops = metrics["comandas_created"] + metrics["debits_requested"]
        ops_sec = total_ops / elapsed if elapsed > 0 else 0
        avg_lat = sum(metrics["latencies"])/len(metrics["latencies"]) if metrics["latencies"] else 0
        
        print("\n" + "="*40)
        print(f"📊 RELATÓRIO DE ESTRESSE ({elapsed:.0f} segundos)")
        print(f"⚡ Throughput: {ops_sec:.2f} transações por segundo")
        print(f"⏱️ Latência Média WS: {avg_lat:.2f} ms")
        print("-" * 40)
        print(f"🎫 Comandas Emitidas:   {metrics['comandas_created']}")
        print(f"💸 Cobranças Tentadas:  {metrics['debits_requested']}")
        print(f"✅ Cobranças Sucesso:   {metrics['debits_success']}")
        print(f"❌ Saldos Insuficientes: {metrics['debits_failed']}")
        print(f"🔌 Erros Lógicos/Rede:  {metrics['errors']}")
        print("="*40 + "\n")

async def main():
    print("🚀 INICIANDO TESTE DE ESTRESSE OUROBOROS\n")
    print("1️⃣ Semeando categorias iniciais...")
    for cat in categories:
        post_json("/api/categories", {"name": cat["name"], "price": cat["price"]}, ADMIN_TOKEN)
        
    print("2️⃣ Registrando pontos de venda (Lojas)...")
    store_tokens = []
    
    # Pega lojas existentes primeiro para evitar duplicação no dashboard, 
    # mas cria novas se for necessário.
    existing_stores_req = urllib.request.Request(f"{BASE_URL}/api/stores", headers={'token': ADMIN_TOKEN})
    try:
        with urllib.request.urlopen(existing_stores_req) as response:
            existing = json.loads(response.read().decode())
            for s in existing:
                store_tokens.append((s["terminal_token"], s["name"]))
    except:
        pass

    # Cria até atingir 5 lojas para o teste
    for i in range(len(store_tokens), 5):
        s_name = store_names[i]
        resp = post_json("/api/stores", {"name": s_name}, ADMIN_TOKEN)
        if resp and "terminal_token" in resp:
            store_tokens.append((resp["terminal_token"], s_name))
            
    if not store_tokens:
        print("❌ Falha crítica: Nenhuma loja criada/recuperada. Verifique se o backend está rodando e aceite o ADMIN_TOKEN.")
        return
        
    print(f"✅ {len(store_tokens)} Lojas preparadas para atacar!\n")
    
    tasks = [
        asyncio.create_task(admin_worker()),
        asyncio.create_task(reporter_loop())
    ]
    
    for token, s_name in store_tokens:
        tasks.append(asyncio.create_task(store_worker(token, s_name)))
        
    print("🔥 DISPARANDO CONEXÕES! (Pressione Ctrl+C para parar)\n")
    try:
        await asyncio.gather(*tasks)
    except asyncio.CancelledError:
        pass

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n\n🛑 TESTE ABORTADO PELO USUÁRIO. FIM.")

File: backend-node/stress_test.js
'use strict';

/**
 * Stress test for the Ouroboros Node.js backend.
 *
 * Usage:
 *   node stress_test.js [duration_seconds=60] [num_stores=5]
 *
 * Requires the server to be running:
 *   npm start   (or npm run dev)
 *
 * What it does:
 *   - 1 admin worker creates comandas at ~2/sec
 *   - N store workers each alternate between balance_query and debit_request
 *   - Reports throughput, latency (p50/p95/p99) and success/failure counts every 3s
 *   - Reconnects automatically on disconnect
 *   - Prints a final summary on Ctrl+C or when duration expires
 */

require('dotenv').config();
const WebSocket = require('ws');

const BASE_URL  = 'http://localhost:8000';
const WS_URL    = 'ws://localhost:8000';
const ADMIN_TOKEN = process.env.ADMIN_TOKEN || 'admin_token_change_me';

const DURATION_SEC    = parseInt(process.argv[2] || '60', 10);
const NUM_STORES      = parseInt(process.argv[3] || '5', 10);
const REPORT_INTERVAL = 3000; // ms

// ---------------------------------------------------------------------------
// Shared state
// ---------------------------------------------------------------------------

const activeCodandas = []; // comanda codes available for debit/query
let running = true;

// ---------------------------------------------------------------------------
// Metrics
// ---------------------------------------------------------------------------

const metrics = {
  comandasCreated:  0,
  debitsConfirmed:  0,
  debitsInsufficient: 0,
  debitsOtherReject:  0,
  balanceQueries:   0,
  errors:           0,
  reconnects:       0,
  latencies:        [],   // all-time, for final report
  windowOps:        0,    // ops completed since last report
  windowStart:      Date.now(),
};

function recordLatency(ms) {
  metrics.latencies.push(ms);
  // cap memory: keep last 2000 samples
  if (metrics.latencies.length > 2000) metrics.latencies.shift();
}

function percentile(sorted, p) {
  if (sorted.length === 0) return 0;
  const idx = Math.ceil((p / 100) * sorted.length) - 1;
  return sorted[Math.max(0, idx)];
}

function latencyStats() {
  if (metrics.latencies.length === 0) return { p50: 0, p95: 0, p99: 0, avg: 0 };
  const sorted = [...metrics.latencies].sort((a, b) => a - b);
  const avg = sorted.reduce((s, v) => s + v, 0) / sorted.length;
  return {
    avg:  avg.toFixed(1),
    p50:  percentile(sorted, 50).toFixed(1),
    p95:  percentile(sorted, 95).toFixed(1),
    p99:  percentile(sorted, 99).toFixed(1),
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function rand(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function pick(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

/** HTTP request helper (no extra deps — uses built-in http module). */
function httpRequest(method, path, body, headers = {}) {
  return new Promise((resolve, reject) => {
    const url = new URL(`${BASE_URL}${path}`);
    const payload = body ? JSON.stringify(body) : null;
    const options = {
      hostname: url.hostname,
      port:     url.port || 8000,
      path:     url.pathname,
      method,
      headers: {
        'Content-Type': 'application/json',
        token: ADMIN_TOKEN,
        ...headers,
        ...(payload ? { 'Content-Length': Buffer.byteLength(payload) } : {}),
      },
    };
    const req = require('http').request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(data) }); }
        catch { resolve({ status: res.statusCode, body: data }); }
      });
    });
    req.on('error', reject);
    if (payload) req.write(payload);
    req.end();
  });
}

/**
 * Connect a WebSocket and wait for the initial 'connected' greeting.
 * Returns { ws, connected }.
 */
function connectWs(url) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(url);
    const timer = setTimeout(() => reject(new Error(`WS connect timeout: ${url}`)), 5000);
    ws.once('open', () => {});
    ws.once('error', (e) => { clearTimeout(timer); reject(e); });
    ws.once('message', (data) => {
      clearTimeout(timer);
      try { resolve({ ws, connected: JSON.parse(data.toString()) }); }
      catch (e) { reject(e); }
    });
  });
}

/**
 * Wait for the next message whose `type` is in the `acceptTypes` set.
 * Silently discards broadcasts (balance_updated, admin_balance_updated,
 * update_next_code) that arrive while waiting for a direct reply.
 */
function waitForType(ws, acceptTypes, timeout = 5000) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error(`Timeout waiting for [${acceptTypes}]`)),
      timeout
    );
    const handler = (data) => {
      let msg;
      try { msg = JSON.parse(data.toString()); }
      catch (e) { clearTimeout(timer); ws.removeListener('message', handler); reject(e); return; }

      if (acceptTypes.includes(msg.type)) {
        clearTimeout(timer);
        ws.removeListener('message', handler);
        resolve(msg);
      }
      // else: broadcast — keep waiting
    };
    ws.on('message', handler);
  });
}

// ---------------------------------------------------------------------------
// Admin worker — creates comandas
// ---------------------------------------------------------------------------

const NAMES = ['Pedro', 'Maria', 'João', 'Ana', 'Lucas', 'Julia', 'Marcos', 'Carla', 'Bruno', 'Lara'];
const CATEGORIES = ['Bolo de Pote', 'Refrigerante', 'Fatia de Pizza', 'Pulseira Neon', 'Hambúrguer'];

async function adminWorker() {
  const url = `${WS_URL}/ws/admin?token=${encodeURIComponent(ADMIN_TOKEN)}`;

  while (running) {
    let ws;
    try {
      ({ ws } = await connectWs(url));

      while (running) {
        const holder = `${pick(NAMES)} ${rand(100, 999)}`;
        const balance = rand(50, 200);
        const cart = [{ name: pick(CATEGORIES), quantity: rand(1, 3) }];

        const t0 = Date.now();
        ws.send(JSON.stringify({
          type: 'create_comanda',
          holder_name: holder,
          initial_balance: balance,
          cart_items: cart,
        }));

        const reply = await waitForType(ws, ['comanda_created', 'error']);
        const latency = Date.now() - t0;

        if (reply.type === 'comanda_created') {
          activeCodandas.push(reply.code);
          // cap list to avoid unbounded growth
          if (activeCodandas.length > 500) activeCodandas.shift();
          metrics.comandasCreated++;
          metrics.windowOps++;
          recordLatency(latency);
        } else {
          metrics.errors++;
        }

        await sleep(500); // ~2 comandas/sec
      }

      ws.close();
    } catch (e) {
      if (running) {
        metrics.errors++;
        metrics.reconnects++;
        await sleep(2000);
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Store worker — debits + balance queries
// ---------------------------------------------------------------------------

async function storeWorker(token, name) {
  const url = `${WS_URL}/ws/store?token=${encodeURIComponent(token)}`;

  while (running) {
    let ws;
    try {
      ({ ws } = await connectWs(url));

      while (running) {
        if (activeCodandas.length === 0) {
          await sleep(200);
          continue;
        }

        const code   = pick(activeCodandas);
        const doDebit = Math.random() < 0.7; // 70% debit, 30% query

        const t0 = Date.now();

        if (doDebit) {
          const amount = rand(5, 40);
          ws.send(JSON.stringify({ type: 'debit_request', comanda_code: code, amount }));
          const reply = await waitForType(ws, ['debit_confirmed', 'debit_rejected']);
          const latency = Date.now() - t0;
          recordLatency(latency);
          metrics.windowOps++;

          if (reply.type === 'debit_confirmed') {
            metrics.debitsConfirmed++;
          } else if (reply.reason === 'insufficient_balance') {
            metrics.debitsInsufficient++;
          } else {
            metrics.debitsOtherReject++;
          }
        } else {
          ws.send(JSON.stringify({ type: 'balance_query', comanda_code: code }));
          await waitForType(ws, ['balance_response', 'error']);
          const latency = Date.now() - t0;
          recordLatency(latency);
          metrics.windowOps++;
          metrics.balanceQueries++;
        }

        await sleep(rand(100, 400));
      }

      ws.close();
    } catch (e) {
      if (running) {
        metrics.errors++;
        metrics.reconnects++;
        await sleep(2000);
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Reporter
// ---------------------------------------------------------------------------

function printReport(elapsed, final = false) {
  const windowSec = (Date.now() - metrics.windowStart) / 1000;
  const throughput = (metrics.windowOps / windowSec).toFixed(1);
  const lat = latencyStats();
  const totalOps = metrics.comandasCreated + metrics.debitsConfirmed +
                   metrics.debitsInsufficient + metrics.debitsOtherReject + metrics.balanceQueries;
  const debitTotal = metrics.debitsConfirmed + metrics.debitsInsufficient + metrics.debitsOtherReject;
  const successRate = debitTotal > 0
    ? ((metrics.debitsConfirmed / debitTotal) * 100).toFixed(1)
    : 'N/A';

  const header = final ? '=== RELATÓRIO FINAL ===' : `=== ${elapsed}s ===`;
  console.log('\n' + '─'.repeat(44));
  console.log(header);
  console.log('─'.repeat(44));
  console.log(`Throughput (janela):  ${throughput} ops/s`);
  console.log(`Latência  avg/p50/p95/p99:  ${lat.avg}/${lat.p50}/${lat.p95}/${lat.p99} ms`);
  console.log('─'.repeat(44));
  console.log(`Comandas emitidas:    ${metrics.comandasCreated}`);
  console.log(`Débitos confirmados:  ${metrics.debitsConfirmed}  (${successRate}% de sucesso)`);
  console.log(`  Saldo insuficiente: ${metrics.debitsInsufficient}`);
  console.log(`  Outros rejeites:    ${metrics.debitsOtherReject}`);
  console.log(`Consultas de saldo:   ${metrics.balanceQueries}`);
  console.log(`Total de operações:   ${totalOps}`);
  console.log(`Erros / reconexões:   ${metrics.errors} / ${metrics.reconnects}`);
  console.log('─'.repeat(44));

  // reset window
  metrics.windowOps = 0;
  metrics.windowStart = Date.now();
}

async function reporterLoop(startTime) {
  while (running) {
    await sleep(REPORT_INTERVAL);
    if (!running) break;
    const elapsed = Math.floor((Date.now() - startTime) / 1000);
    printReport(elapsed);
  }
}

// ---------------------------------------------------------------------------
// Setup — seed categories and create/reuse stores
// ---------------------------------------------------------------------------

async function setup() {
  console.log('\nIniciando teste de estresse do backend-node...');
  console.log(`Duração: ${DURATION_SEC}s | Lojas: ${NUM_STORES}\n`);

  // Seed categories
  process.stdout.write('Semeando categorias... ');
  for (const name of CATEGORIES) {
    await httpRequest('POST', '/api/categories', { name, price: rand(5, 30) });
  }
  console.log('ok');

  // Fetch or create stores
  process.stdout.write('Preparando lojas... ');
  const { body: existing } = await httpRequest('GET', '/api/stores');
  const storeTokens = Array.isArray(existing)
    ? existing.map((s) => ({ token: s.terminal_token, name: s.name }))
    : [];

  const storeNames = ['Cantina Central', 'Bazar 3º Ano', 'Barraca de Jogos', 'Doces e Cia', 'Artesanato'];
  while (storeTokens.length < NUM_STORES) {
    const name = storeNames[storeTokens.length] || `Loja ${storeTokens.length + 1}`;
    const { body } = await httpRequest('POST', '/api/stores', { name });
    if (body && body.terminal_token) {
      storeTokens.push({ token: body.terminal_token, name });
    } else {
      console.error('\nFalha ao criar loja. Servidor rodando? ADMIN_TOKEN correto?');
      process.exit(1);
    }
  }
  console.log(`${storeTokens.length} lojas prontas`);
  return storeTokens.slice(0, NUM_STORES);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const stores = await setup();
  const startTime = Date.now();

  console.log('\nDisparando workers... (Ctrl+C para parar)\n');

  const tasks = [
    adminWorker(),
    reporterLoop(startTime),
    ...stores.map(({ token, name }) => storeWorker(token, name)),
  ];

  // Stop after DURATION_SEC
  const deadline = sleep(DURATION_SEC * 1000).then(() => {
    running = false;
  });

  process.on('SIGINT', () => {
    running = false;
  });

  await Promise.race([deadline, ...tasks.map((t) => t.catch(() => {}))]);
  running = false;

  // Give workers a moment to finish current ops
  await sleep(500);

  const elapsed = Math.floor((Date.now() - startTime) / 1000);
  printReport(elapsed, true);
  process.exit(0);
}

main().catch((e) => {
  console.error('Erro fatal:', e.message);
  process.exit(1);
});
